lib/_http_client.js

```diff
@@ -337,25 +339,47 @@ ClientRequest.prototype._implicitHeader = function _implicitHeader() {
 };
 
 ClientRequest.prototype.abort = function abort() {
-  if (!this.aborted) {
-    process.nextTick(emitAbortNT, this);
+  if (this.aborted) {
+    return;
   }
   this.aborted = true;
+  process.nextTick(emitAbortNT, this);
+  this.destroy();
+};
+
+ClientRequest.prototype.destroy = function destroy(err) {
+  if (this.destroyed) {
+    return;
+  }
+  this.destroyed = true;
 
   // If we're aborting, we don't care about any more response data.
   if (this.res) {
     this.res._dump();
   }
 
-  // In the event that we don't have a socket, we will pop out of
-  // the request queue through handling in onSocket.
-  if (this.socket) {
-    // in-progress
-    this.socket.destroy();
+  if (!this.socket) {
+    this.once(kOnSocket, (socket) => {
```

Comment:
Instead of adding yet another listener, wouldn't it be better to create another function named `cleanup` and call it here? And instead of [line 715](https://github.com/nodejs/node/pull/32148/files#diff-e3bc37430eb078ccbafe3aa3b570c91aR715) it would be

```js
  if (this.destroyed) {
    cleanup(this);
    return;
  }
```

It's just an idea. I wouldn't use event emitters if possible. They tend to use much more memory and provide inconsistent (or unpredictable) behavior if called multiple times.

Reply:
I've updated the PR with a version that is closer to the original, so this is not needed. PTAL.